sdk/opendp/whitenoise/data/dpsu.py

```diff
@@ -0,0 +1,117 @@
+import pandas as pd
+import math
+import random
+import operator
+from collections import defaultdict
+
+from opendp.whitenoise.metadata import CollectionMetadata
+from opendp.whitenoise.sql import PandasReader
+from opendp.whitenoise.sql.parse import QueryParser
+from opendp.whitenoise.mechanisms.rand import laplace
+
+def preprocess_df_from_query(meta_path, csv_path, query, table_name):
+	"""
+	Returns a dataframe with user_id | tuple based on query grouping keys.
+	"""
+	schema = CollectionMetadata.from_file(meta_path)
+	df = pd.read_csv(csv_path)
+
+	reader = PandasReader(schema, df)
+	queries = QueryParser(schema).queries(query)
+	query_ast = queries[0]
+	
+	group_cols = [ge.expression.name for ge in query_ast.agg.groupingExpressions]
+	kc = schema[table_name].key_cols()[0].name
+
+	preprocessed_df = pd.DataFrame()
+	preprocessed_df[kc] = df[kc]
+	preprocessed_df["group_cols"] = tuple(df[group_cols].values.tolist())
+
+	return preprocessed_df
+
+def reservoir_sample(iterable, max_contrib):
```

Comment:
The sample really just needs to be a uniform sample without replacement.  The reservoir sampling approach is a way to obtain a uniform sample without replacement when the number of elements to choose from is unknown ahead of time and unlikely to fit in memory.  Since you have all of the elements in the grouping anyway, you can just randomly do something like np.choose.

I forgot to mention the source of randomness should be cryptographic, though, so np.choose won't be random enough.  Another way to get a uniform random sample without replacement is to shuffle the set and grab the first 'k' elements.  But again it is important in that case to shuffle in a way that is cryptographically random rather than using a PRNG.

I believe that is very low risk in this case, FWIW, since I think the shuffling by user, which is accomplished up front via the hash, is more important for this algorithm.  But we should do all of them with strong randomness